### PR TITLE
[megatron, hardware] fix: pure-torch fast_hadamard_transform fallback for DSA on ROCm

### PR DESCRIPTION
**Pull request by PeterYang12**
DeepSeek sparse-attention (DSA) in Megatron-LM binds `from fast_hadamard_transform import hadamard_transform` at import time and asserts it is not None inside the indexer's `rotate_activation`. The Dao-AILab package only ships nvcc kernels and its setup.py hard-requires nvcc, so it cannot be built on ROCm: there the import yields None and every DSA forward crashes.

Register a vectorized Fast Walsh-Hadamard Transform under the real import name so any importer picks it up, and back-fill modules that already captured None. The back-fill is required, not defensive: Megatron's dsa.py binds the symbol at import time, and the mbridge path in megatron_utils.get_model() calls apply_patch() after the model has been built. No-op when the real CUDA package is installed.

Split out of #7050 so it can be reviewed independently of the ROCm Dockerfile and FP8/rollout changes there.

### What does this PR do?

> Add **concise** overview of what this PR aims to achieve or accomplish. Reference related GitHub issues and PRs that help with the review.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: ...
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `fully_async`, `one_step_off`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/verl-project/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [ ] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

**Review by wuxibin89 — APPROVED** [review]

File: verl/models/mcore/patch.py [hunk 1]
```diff
@@ -17,7 +17,89 @@
...
+    """
+    import logging
+    import sys
+    import types
+
+    import torch
+
+    # Real CUDA package present and working -> leave it alone.
+    existing = sys.modules.get("fast_hadamard_transform")
+    if existing is not None and getattr(existing, "hadamard_transform", None) is not None:
```
Comment (wuxibin89) [inline verl/models/mcore/patch.py]: This predictor is not right.

**Review by wuxibin89 — CHANGES_REQUESTED** [review]

File: verl/models/mcore/patch.py (hunk 1 above)
Comment (PeterYang12) [reply verl/models/mcore/patch.py]: Refactor these part codes and use importlib to detect the package installation.
My UT test on H200, test pure pytorch implementation and CUDA version.
image
test codes:
[test_mcore_hadamard_transform.py](https://github.com/user-attachments/files/30691800/test_mcore_hadamard_transform.py)

File: verl/models/mcore/patch.py [hunk 2]
```diff
@@ -17,7 +17,98 @@
...
+    import sys
+    import types
+
+    # Import rather than probe ``sys.modules``: an entry under that name only means
+    # *something* is registered (a partially initialized module, a stub from another
+    # framework, an earlier run of this shim), which says nothing about whether a
+    # usable kernel is there. A broken CUDA extension can also fail the import with
+    # OSError/RuntimeError, not just ImportError.
+    try:
+        module = importlib.import_module("fast_hadamard_transform")
```
Comment (wuxibin89) [inline verl/models/mcore/patch.py]: Return here if fast_hadamard_transform exists?
